### PR TITLE
generalized most SRM unit tests to None and Compiled

### PR DESCRIPTION
**Pull request by veanes**
Cleaned up (most) of the unit tests that used to be specific for `NonBacktracking` in `RegexSRMTests` by generalizing them to both `None` and `Compiled` options. 
Added a new test called `AllMatches` in `Regex.Match.Tests` that consolidates most prior SRM specific tests and lifts **all these test also to None and Compiled** (thus further checking identical results in may corner cases -- that were actually not previously formally tested)
Also lifted over some other tests. 
Some further cleanup may still be needed for SRM specific tests but this was the bulk of them.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs [hunk 1]
```diff
@@ -647,13 +663,20 @@ public void Docs_Backtracking_WithOptionalQuantifiersOrAlternationConstructs(Reg
...
-            Assert.False(Regex.IsMatch(Input, "^((?>[0-9a-fA-F]{1,4}:)*(?>[0-9a-fA-F]{1,4}))*(::)$", options)); // non-backtracking
+            if (options == RegexHelpers.RegexOptionNonBacktracking)
+            {
+                // With NonBacktracking option this test becomes feasible: from almost a minute to less than a second
+                Assert.False(Regex.IsMatch(Input, "^(([0-9a-fA-F]{1,4}:)*([0-9a-fA-F]{1,4}))*(::)$", options));
+            }
+            else
+            {
+                Assert.False(Regex.IsMatch(Input, "^((?>[0-9a-fA-F]{1,4}:)*(?>[0-9a-fA-F]{1,4}))*(::)$", options)); // forced non-backtracking
+            }
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs:660]:
```suggestion
            Assert.False(Regex.IsMatch(Input, "^((?>[0-9a-fA-F]{1,4}:)*(?>[0-9a-fA-F]{1,4}))*(::)$", options)); // forced non-backtracking
```

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs [hunk 2]
```diff
@@ -647,13 +663,20 @@ public void Docs_Backtracking_WithOptionalQuantifiersOrAlternationConstructs(Reg
 
         // https://docs.microsoft.com/en-us/dotnet/standard/base-types/backtracking-in-regular-expressions#nonbacktracking-subexpression
         [Theory]
-        [InlineData(RegexOptions.None)]
-        [InlineData(RegexOptions.Compiled)]
+        [MemberData(nameof(RegexOptions_TestData))]
         public void Docs_Backtracking_WithNestedOptionalQuantifiers(RegexOptions options)
         {
             const string Input = "b51:4:1DB:9EE1:5:27d60:f44:D4:cd:E:5:0A5:4a:D24:41Ad:";
             // Assert.False(Regex.IsMatch(Input, "^(([0-9a-fA-F]{1,4}:)*([0-9a-fA-F]{1,4}))*(::)$")); // takes too long due to backtracking
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs]:
```suggestion
```

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs [hunk 3]
```diff
@@ -1312,5 +1312,166 @@ public void Match_Wordchar()
...
+            Assert.True(re.IsMatch(input));
+        }
+
+        public static IEnumerable<object[]> AllMatches_TestData()
+        {
+            foreach (RegexOptions options in new[] { RegexHelpers.RegexOptionNonBacktracking, RegexOptions.None, RegexOptions.Compiled })
+            {
+                if (PlatformDetection.IsNetFramework && options == RegexHelpers.RegexOptionNonBacktracking)
+                {
+                    yield break;
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: Did you really mean `yield break;` here?  Shouldn't this be:
```suggestion
                    continue;
```
?

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs [hunk 4]
```diff
@@ -1312,5 +1312,166 @@ public void Match_Wordchar()
...
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.Compiled)]
+        [InlineData(RegexHelpers.RegexOptionNonBacktracking)]
+        public void StressTestDeepNestingOfConcat(RegexOptions options)
+        {
+            if (PlatformDetection.IsNetFramework && options == RegexHelpers.RegexOptionNonBacktracking)
+            {
+                return;
+            }
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: Could we move the RegexOptionsExtended and RegexOptions_TestData into a shared test class (I believe we already have one where we put helpers)?  Then this can just use that same MemberData, and we can get rid of some duplication elsewhere as well.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs [hunk 5]
```diff
@@ -1312,5 +1312,166 @@ public void Match_Wordchar()
...
+
+            string pattern = string.Concat(Enumerable.Repeat("([a-z]", 1000).Concat(Enumerable.Repeat(")", 1000)));
+            string input = string.Concat(Enumerable.Repeat("abcde", 200));
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: Nit: could you lift these constant values out into named const locals?  I'm also wondering if 1000 is sufficient, especially on Linux where stacks are typically much larger.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 4 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: It turns out that `MemberData` can only reference a method in the same class where the test itself is defined, so it seems moving `RegexOptions_TestData` to `RegexHelpers` class is not allowed. I'll move `RegexOptionsExtended`.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs (hunk 1 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs:660]: The atomic loop annotation  `(?>...)` is not supported in `NonBacktracking` mode. So the `IsMatch` will throw an exception if not guarded by `else`.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 3 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: I'm updating the `Regex.Match.Tests` to use the new `RegexHelpers.RegexOptionsExtended()` method instead to clean this up, whenever possible. The intent here was simply that the tests must not run in the `NonBacktracking` mode when `IsNetFramework` is true.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: > It turns out that MemberData can only reference a method in the same class where the test itself is defined

You can set a MemberType property on the attribute, e.g.
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Linq.Parallel/tests/QueryOperators/OfTypeTests.cs#L33

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: > The intent here was simply that the tests must not run in the NonBacktracking mode when IsNetFramework is true.

Right, but `yield break;` will permanently exit the enumeration.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs:660]: Ah, ok.  Can you remind me why?  Rather than writing it one-off here, I think we should add a README.md to the root of the SRM folder that outlines:
- What features aren't supported (and why)?
- Key terminology employed in the implementation
- Other interesting details to help understand the implementation

We can do that subsequently.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 5 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: I introduced `RegexHelpers.StressTestNestingDepth` and set it  to 10000. I'm getting 
`The program '[55724] dotnet.exe' has exited with code 3221225477 (0xc0000005) 'Access violation'.`
so need to investigate this further.

Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: This happens only in `NonBacktracking` mode, the other cases seem fine.

Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: I'm investigating this next, but will set it back to 1000 for now to finish these edits first,

File: src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs (hunk 1 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs:660]: Short reply is that it changes the meaning of the loop -- cutting of choices that would exist with backtracking. In `NonBacktracking` mode it messes with/contradicts the DFA semantics because all search paths are preserved there -- just that backtracking is not being used to preserve those paths. We discussed this briefly at some point in the spring. Yes, **documentation** is the next big item on the todo list.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 4 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: Good to know. Thanks! I'll use that attribute.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 3 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: This is not what I intended, should be fixed now with the commit I'm pushing next.

**Comment by stephentoub** [comment]
(Pushed a rebase to handle conflicts with other changes)

**Review by stephentoub — APPROVED** [review]
Thanks.
